Allow to set a default expiration value on the generated token

dunglas opened the pull request:
Generating JWT without expiration is a bad security practice. This PR adds the ability to set a default lifetime for generated tokens using the new `Authorization` util.

azjezz commented on src/Publisher.php:
```diff
@@ -37,6 +37,7 @@ final class Publisher implements PublisherInterface
 
     /**
      * @param TokenProviderInterface|callable(Update $update):string $jwtProvider
+     * @param mixed                                  $jwtProvider
```
```suggestion
     * @param (callalbe(Update): string)|ToeknProviderInterface $jwtProvider
```

dunglas replied:
This breaks PHP CS Fixer

mtarld commented on src/Authorization.php:
```diff
@@ -23,10 +23,15 @@ final class Authorization
+    /**
+     * @param int|null $jwtLifetime If not null, an "exp" claim is always set to now + $jwtLifetime (in seconds)
+     */
+    public function __construct(HubRegistry $registry, int $jwtLifetime = null)
```
```suggestion
    public function __construct(HubRegistry $registry, ?int $jwtLifetime = null)
```
For consistency?

dunglas replied:
We can even set a default value then, it will be safer.

mtarld commented on tests/AuthorizationTest.php:
```diff
@@ -0,0 +1,46 @@
+            new StaticTokenProvider('foo.bar.baz'),
+            function (Update $u): string { return 'dummy'; },
+            new LcobucciFactory('secret')
+        ), []);
```
```suggestion
        ));
```
Not needed anymore

chalasr commented:
👍 For the record, I think this is the only standard claim that should be set with a good default by this library.
I mean, after the recent work done on the authorization part, people will eventually start asking for `iss` and other claims to be set automatically, then ask for a configurable clock skew, ... endless story. 
The answer should remain the same as before to me: decorate the existing token provider/factory on your own, the system is flexible and pluggable enough.

dunglas commented:
I changed how the `exp` claim is generated, and added a way to control the cookie expiration time accordingly:

> the lifetime of the cookie, the "exp" claim of the JWT will be set accordingly, set to null to use the default value and to 0 to set a session cookie (the default expiration time of the JWT will be 1 hour)

nicolas-grekas commented on src/Jwt/LcobucciFactory.php:
```diff
@@ -37,8 +37,12 @@ final class LcobucciFactory implements TokenFactoryInterface
+    /**
+     * @param int|null $jwtLifetime If not null, an "exp" claim is always set to now + $jwtLifetime (in seconds)
+     */
+    public function __construct(string $secret, string $algorithm = 'hmac.sha256', ?int $jwtLifetime = 3600)
```
why hardcode 3600 instead of reading `ini_get('session.cookie_lifetime')`?

dunglas replied:
Good idea. Fixed. I just added a fallback to 3600 in case the value is set to "session", as this concept doesn't exist for JWT.
